In [1]:
import pandas as pd

df = pd.read_csv("COVID-data_consolidated - County_City Race 1.csv", header = 0)
df.describe()
df.head()
df

date     county     city           race  population  \
0     2021-03-09    Suffolk  Chelsea          AI/AN        53.0   
1     2021-03-09    Suffolk  Chelsea          Asian       810.0   
2     2021-03-09    Suffolk  Chelsea          Black      3336.0   
3     2021-03-09    Suffolk  Chelsea       Hispanic     25092.0   
4     2021-03-09    Suffolk  Chelsea          Multi       420.0   
...          ...        ...      ...            ...         ...   
4477  2022-10-04  Middlesex  Everett          Multi      4581.0   
4478  2022-10-04  Middlesex  Everett          NH/PI        15.0   
4479  2022-10-04  Middlesex  Everett          White     19034.0   
4480  2022-10-04  Middlesex  Everett  Other/Unknown         0.0   
4481  2022-10-04  Middlesex  Everett          Total     49075.0   

      atleast_onedose  fully_vaccinated  boosted  boosted_second  \
0                 0.0               0.0      NaN             NaN   
1               207.0             109.0      NaN             NaN   
2               399.0             237.0      NaN             NaN   
3              1791.0             810.0      NaN             NaN   
4                81.0              50.0      NaN             NaN   
...               ...               ...      ...             ...   
4477           1190.0            1135.0    699.0           129.0   
4478              0.0               0.0      0.0             0.0   
4479          15125.0           13465.0   7679.0          1956.0   
4480              0.0               0.0      0.0             0.0   
4481          44716.0           37524.0  18074.0          3200.0   

      partially_vaccinated  ...  Unnamed: 16  Unnamed: 17  Unnamed: 18  \
0                      0.0  ...          NaN          NaN          NaN   
1                     98.0  ...          NaN          NaN          NaN   
2                    162.0  ...          NaN          NaN          NaN   
3                    981.0  ...          NaN          NaN          NaN   
4                     31.0  ...          NaN          NaN          NaN   
...                    ...  ...          ...          ...          ...   
4477                   NaN  ...          NaN          NaN          NaN   
4478                   NaN  ...          NaN          NaN          NaN   
4479                   NaN  ...          NaN          NaN          NaN   
4480                   NaN  ...          NaN          NaN          NaN   
4481                   NaN  ...          NaN          NaN          NaN   

      Unnamed: 19     Unnamed: 20  Unnamed: 21  fully/population  \
0             NaN      2021-04-06      chelsea          0.057587   
1             NaN       hispanic        newton          0.007940   
2             NaN  last four cols       revere          0.035153   
3             NaN             NaN     wellsley          0.004140   
4             NaN             NaN          NaN               NaN   
...           ...             ...          ...               ...   
4477          NaN             NaN          NaN               NaN   
4478          NaN             NaN          NaN               NaN   
4479          NaN             NaN          NaN               NaN   
4480          NaN             NaN          NaN               NaN   
4481          NaN             NaN          NaN               NaN   

      partially/population  fully/total  partially/total  
0                 0.060350     0.333177         0.455810  
1                 0.007525     0.026112         0.041788  
2                 0.044241     0.184190         0.303358  
3                 0.004880     0.018996         0.029364  
4                      NaN          NaN              NaN  
...                    ...          ...              ...  
4477                   NaN          NaN              NaN  
4478                   NaN          NaN              NaN  
4479                   NaN          NaN              NaN  
4480                   NaN          NaN              NaN  
4481                   NaN         

In [2]:
df['percentage_fully_vaccinated'] = df['fully_vaccinated']/df['population']

In [3]:
df

date     county     city           race  population  \
0     2021-03-09    Suffolk  Chelsea          AI/AN        53.0   
1     2021-03-09    Suffolk  Chelsea          Asian       810.0   
2     2021-03-09    Suffolk  Chelsea          Black      3336.0   
3     2021-03-09    Suffolk  Chelsea       Hispanic     25092.0   
4     2021-03-09    Suffolk  Chelsea          Multi       420.0   
...          ...        ...      ...            ...         ...   
4477  2022-10-04  Middlesex  Everett          Multi      4581.0   
4478  2022-10-04  Middlesex  Everett          NH/PI        15.0   
4479  2022-10-04  Middlesex  Everett          White     19034.0   
4480  2022-10-04  Middlesex  Everett  Other/Unknown         0.0   
4481  2022-10-04  Middlesex  Everett          Total     49075.0   

      atleast_onedose  fully_vaccinated  boosted  boosted_second  \
0                 0.0               0.0      NaN             NaN   
1               207.0             109.0      NaN             NaN   
2               399.0             237.0      NaN             NaN   
3              1791.0             810.0      NaN             NaN   
4                81.0              50.0      NaN             NaN   
...               ...               ...      ...             ...   
4477           1190.0            1135.0    699.0           129.0   
4478              0.0               0.0      0.0             0.0   
4479          15125.0           13465.0   7679.0          1956.0   
4480              0.0               0.0      0.0             0.0   
4481          44716.0           37524.0  18074.0          3200.0   

      partially_vaccinated  ...  Unnamed: 17  Unnamed: 18  Unnamed: 19  \
0                      0.0  ...          NaN          NaN          NaN   
1                     98.0  ...          NaN          NaN          NaN   
2                    162.0  ...          NaN          NaN          NaN   
3                    981.0  ...          NaN          NaN          NaN   
4                     31.0  ...          NaN          NaN          NaN   
...                    ...  ...          ...          ...          ...   
4477                   NaN  ...          NaN          NaN          NaN   
4478                   NaN  ...          NaN          NaN          NaN   
4479                   NaN  ...          NaN          NaN          NaN   
4480                   NaN  ...          NaN          NaN          NaN   
4481                   NaN  ...          NaN          NaN          NaN   

         Unnamed: 20  Unnamed: 21  fully/population  partially/population  \
0         2021-04-06      chelsea          0.057587              0.060350   
1          hispanic        newton          0.007940              0.007525   
2     last four cols       revere          0.035153              0.044241   
3                NaN     wellsley          0.004140              0.004880   
4                NaN          NaN               NaN                   NaN   
...              ...          ...               ...                   ...   
4477             NaN          NaN               NaN                   NaN   
4478             NaN          NaN               NaN                   NaN   
4479             NaN          NaN               NaN                   NaN   
4480             NaN          NaN               NaN                   NaN   
4481             NaN          NaN               NaN                   NaN   

      fully/total  partially/total  percentage_fully_vaccinated  
0        0.333177         0.455810                     0.000000  
1        0.026112         0.041788                     0.134568  
2        0.184190         0.303358                     0.071043  
3        0.018996         0.029364                     0.032281  
4             NaN              NaN                     0.119048  
...           ...              ...                          ...  
4477          NaN              NaN                     0.247762  
4478          NaN              NaN              

## Comparing percentage of fully vaccinated population

In [5]:
from scipy.stats import ttest_ind

city = ['Chelsea', 'Newton', 'Revere', 'Wellesley','Springfield', 'Everett']
race = ['Black','Hispanic','White','Asian']

def t_test(df1, df2):
	stat,p = ttest_ind(df1, df2 ,equal_var = False)
	print('stat=%.3f, p=%.8f' % (stat, p))
	if p > 0.05:
		print('Probably the same distribution')
	else:
		print('Probably different distributions')

	return p


def hypothesis_test(df):
	for i in range(len(city)-1):
		for j in range(i+1, len(city)):
			df1 = df[df['city'] == city[i]]
			df2 = df[df['city'] == city[j]]
			print(city[i], 'vs.', city[j])
			for a in range(len(race)):
				df_1 = df1[df1['race'] == race[a]]
				df_2 = df2[df2['race'] == race[a]]
				print(race[a])
				t_test(df_1['percentage_fully_vaccinated'], df_2['percentage_fully_vaccinated'])

			print('\n')

hypothesis_test(df)

Chelsea vs. Newton
Black
stat=-10.864, p=0.00000000
Probably different distributions
Hispanic
stat=-0.113, p=0.91052820
Probably the same distribution
White
stat=1.601, p=0.11119134
Probably the same distribution
Asian
stat=9.869, p=0.00000000
Probably different distributions


Chelsea vs. Revere
Black
stat=-0.725, p=0.46959120
Probably the same distribution
Hispanic
stat=0.622, p=0.53475560
Probably the same distribution
White
stat=7.675, p=0.00000000
Probably different distributions
Asian
stat=10.377, p=0.00000000
Probably different distributions


Chelsea vs. Wellesley
Black
stat=-2.779, p=0.00611275
Probably different distributions
Hispanic
stat=7.462, p=0.00000000
Probably different distributions
White
stat=4.463, p=0.00001497
Probably different distributions
Asian
stat=13.896, p=0.00000000
Probably different distributions


Chelsea vs. Springfield
Black
stat=1.593, p=0.11317241
Probably the same distribution
Hispanic
stat=9.148, p=0.00000000
Probably different distributions
White

### Summary


Chelsea vs. Newton
Black
stat=-10.864, p=0.00000000
Probably different distributions

Hispanic
stat=-0.113, p=0.91052820
Probably the same distribution

White
stat=1.601, p=0.11119134
Probably the same distribution

Asian
stat=9.869, p=0.00000000
Probably different distributions


When comparing Chelsea’s race data with Newton’s race data, we found that the Hispanic populations and Caucasian populations are statistically similar. Performing the T test to compare the hispanic population data between these two cities results in a p value of 0.910, meaning that these two data distributions are very statistically similar.

Comparing the caucasian population data results in a p value of 0.111. While being above 0.05 (which still proves significance), the p value is substantially lower than the p value of the T test for hispanic data.


Chelsea vs. Revere
Black
stat=-0.725, p=0.46959120
Probably the same distribution
Hispanic
stat=0.622, p=0.53475560
Probably the same distribution
White
stat=7.675, p=0.00000000
Probably different distributions
Asian
stat=10.377, p=0.00000000
Probably different distributions

There also appears to be some similarity between Chelsea and Revere’s data, as comparing their black populations with T test results in a p value of 0.469 and comparing their hispanic populations results in a p value of 0.534.


Chelsea vs. Springfield
Black
stat=1.593, p=0.11317241
Probably the same distribution
Hispanic
stat=9.148, p=0.00000000
Probably different distributions
White
stat=6.618, p=0.00000000
Probably different distributions
Asian
stat=13.536, p=0.00000000
Probably different distributions

Chelsea and Springfield’s black population also show signs of similarity with a T test comparison resulting in a p value of 0.113.



# Rate of Change

In [4]:
from scipy.stats import ttest_ind

city = ['Chelsea', 'Newton', 'Revere', 'Wellesley','Springfield', 'Everett']
race = ['Black','Hispanic','White','Asian']

def t_test(df1, df2):
	stat,p = ttest_ind(df1, df2 ,equal_var = False)
	print('stat=%.3f, p=%.8f' % (stat, p))
	if p > 0.05:
		print('Probably the same distribution')
	else:
		print('Probably different distributions')

	return p


def hypothesis_test(df):
	for i in range(len(city)-1):
		for j in range(i+1, len(city)):
			df1 = df[df['city'] == city[i]]
			df2 = df[df['city'] == city[j]]
			print(city[i], 'vs.', city[j])
			for a in range(len(race)):
				df_1 = df1[df1['race'] == race[a]]
				df_2 = df2[df2['race'] == race[a]]
				diff1 = df_1['percentage_fully_vaccinated'].diff().dropna()
				diff2 = df_2['percentage_fully_vaccinated'].diff().dropna()
				print(race[a])
				t_test(diff1, diff2)

			print('\n')

hypothesis_test(df)

Chelsea vs. Newton
Black
stat=-0.848, p=0.39819699
Probably the same distribution
Hispanic
stat=-0.504, p=0.61484125
Probably the same distribution
White
stat=-0.577, p=0.56468331
Probably the same distribution
Asian
stat=-0.065, p=0.94816446
Probably the same distribution


Chelsea vs. Revere
Black
stat=0.344, p=0.73133018
Probably the same distribution
Hispanic
stat=-0.256, p=0.79806534
Probably the same distribution
White
stat=0.088, p=0.93027658
Probably the same distribution
Asian
stat=-0.015, p=0.98801333
Probably the same distribution


Chelsea vs. Wellesley
Black
stat=0.760, p=0.44888363
Probably the same distribution
Hispanic
stat=1.255, p=0.21145199
Probably the same distribution
White
stat=-0.397, p=0.69213862
Probably the same distribution
Asian
stat=0.112, p=0.91115461
Probably the same distribution


Chelsea vs. Springfield
Black
stat=0.313, p=0.75467528
Probably the same distribution
Hispanic
stat=1.704, p=0.09118182
Probably the same distribution
White
stat=0.054, p=0.9

### Summary


We attempted to figure out if there is a significant difference in rate of change of fully vaccinated population. We took the difference of percentage of the fully vaccinated population between two consecutive times and then used hypothesis testing to examine how similar they are.

When comparing Chelsea and other cities, surprisingly we found that the rates of change for hispanic, black, asisn and white population are all statistically significant because their p-value are greater than the significance level (0.05).
